### ISSUE TITLE
Cannot create LSTM training data from scratch following the wiki without pre-existing trained model: fails with "Tesseract couldn't load any languages!"

### ISSUE THREAD

**Issue by juliangilbey** [tesseract-ocr/tesseract#2616]
### Environment

* **Tesseract Version**: 4.1.0
* **Commit Number**: official 4.1.0 source code
* **Platform**: Linux erdos 4.19.0-5-amd64 #1 SMP Debian 4.19.37-6 (2019-07-18) x86_64 GNU/Linux

### Current Behavior:

tl;dr: following the wiki instructions, use `combine_lang_model` to create a minimal traineddata file, then use `tesstrain.sh` to create initial training data before training a new neural network to perform the OCR.  This fails.

Details:

I am trying to train tesseract from scratch, as I am trying different network structures.  Working in the directory `/tmp/tesstrain`, I cloned the `langdata_lstm` repository into `../langdata_lstm`, and created directories called `tessdata/eng` and `engtrain`.  I found that I needed to also copy the upstream file `tessdata/configs/lstm.train` into `tessdata/eng/configs` (this last step is not mentioned on the wiki, but is needed otherwise `tesstrain.sh` fails on `phase_E_extract_features`).

I then created a starter training file, as described on the wiki, using the command:

```
combine_lang_model --output_dir ./tessdata --pass_through_recoder \
  --input_unicharset ../langdata_lstm/eng/eng.unicharset --script_dir ../langdata_lstm \
  --lang eng --words ../langdata_lstm/eng/eng.wordlist \
  --puncs ../langdata_lstm/eng/eng.punc --numbers ../langdata_lstm/eng/eng.numbers
```

This seemed to work fine, and produced three files in `tessdata/eng`, namely `eng.charset_size=112.txt`, `eng.traineddata`, and `eng.unicharset`.

Now the `eng.traineddata` only has 6 components in it:
TESSDATA_LSTM_NUMBER_DAWG, TESSDATA_LSTM_PUNC_DAWG, TESSDATA_LSTM_RECODER, TESSDATA_LSTM_SYSTEM_DAWG, TESSDATA_LSTM_UNICHARSET and TESSDATA_VERSION (which is "4.0.0").

So far, so good.

I then ran tesstrain.sh to build the initial training data, as on the wiki:

```
/usr/share/tesseract-ocr/tesstrain.sh --fonts_dir /usr/share/fonts --lang eng --linedata_only \
  --noextract_font_properties --fontlist Arial --langdata_dir ../langdata_lstm \
  --tessdata_dir ./tessdata/eng --output_dir ./engtrain 
```

(For testing purposes, I reduced the `eng.training_text` file to a 500 line file, and only used one font.)

Phases "I" and "UP" work fine, as far as I can tell.  But phase_E_extract_features blows up:

```
=== Phase E: Generating lstmf files ===
Using TESSDATA_PREFIX=./tessdata/eng
[Tue 13 Aug 14:51:28 BST 2019] /usr/bin/tesseract /tmp/eng-2019-08-13.pnN/eng.Arial.exp0.tif /tmp/eng-2019-08-13.pnN/eng.Arial.exp0 --psm 6 lstm.train
Failed loading language 'eng'
Tesseract couldn't load any languages!
Could not initialize tesseract.
ERROR: Program tesseract failed. Abort.
```

After some poking around, I figured out at least part of the cause.  Since this call to tesseract is using the minimal training data file in `./tessdata/eng/eng.traineddata`, which does not have a TESSDATA_LSTM component, when it gets into the function `init_tesseract_lang_data` in `src/ccmain/tessedit.cpp`, it switches from OEM_DEFAULT into OEM_TESSERACT_ONLY mode.  (And using `-oem 1` on the tesseract command line fails, too.)  But then it tries to look for a TESSDATA_UNICHARSET in the traineddata file, which doesn't exist (in the test `if (!mgr->GetComponent(TESSDATA_UNICHARSET, &fp) ...)`), so it bombs out.

This means that it seems to be impossible to use `tesstrain.sh` to build a training set from scratch for a new type of network.

(Note that this does look similar to bug #1904, but that was closed by the submitter without any comments.)

### Expected Behavior:

Ideally, `tesstrain.sh` would work out of the box for training from scratch as described on the wiki.

### Suggested Fix:

I don't yet understand tesseract well enough to know whether this would work, but it might be that `tesstrain.sh` is trying to do two different things for LSTM networks: create some training data (images and ground truths, etc.) and incorporate it into the eng.traineddata file, but also to do some initial learning on it (in the step in phase_E which calls tesseract).  If I'm right (which I may well not be), it might be possible to split these steps into two, or to have a command-line option to instruct tesseract that it should not be expecting any TESSERACT_LSTM component to be present, for example.  But I am quite stumped at this point!

Many thanks for all of your superb work on tesseract!

**Comment by juliangilbey** [comment]
Woo-hoo, I've found a trivial patch which fixes this!

Simply add the line:

```
tessedit_init_config_only T
```

to `lstm.train` so that it doesn't actually try building the recognition model.

This is still present in the current HEAD (423a18851348c6729235f383c1be6c529478dc58).

(And the wiki should still mention that `lstm.train` needs to be copied to the configs directory.)

**Comment by amitdo** [comment]
>tl;dr: following the wiki instructions, use combine_lang_model to create a minimal traineddata file, then use tesstrain.sh to create initial training data before training a new neural network to perform the OCR. This fails.

tesstrain.sh runs  combine_lang_model, so you do not need to run it manually.

https://github.com/tesseract-ocr/tesseract/blob/a2b13b49ffffe7ff46498080e5286ef56e2388f1/src/training/tesstrain_utils.sh#L567

**Comment by juliangilbey** [comment]
Unfortunately, that doesn't help, because this command appears in make__lstmdata, which runs after phase_E_extract_features (see tesstrain.sh), and phase_E_extract_features already requires a prepared traineddata file to run.  Here's the output of running tesstrain.sh without a traineddata file already prepared:

```
erdos:/tmp/tesstrain $ /usr/share/tesseract-ocr/tesstrain.sh --fonts_dir /usr/share/fonts --lang eng --linedata_only --noextract_font_properties --fontlist Arial --langdata_dir ../langdata_lstm --tessdata_dir ./tessdata/eng --output_dir ./engtrain 

=== Starting training for language 'eng'
[Fri 16 Aug 00:04:25 BST 2019] /usr/bin/text2image --fonts_dir=/usr/share/fonts --font=Arial --outputbase=/tmp/font_tmp.KohYfRN1r2/sample_text.txt --text=/tmp/font_tmp.KohYfRN1r2/sample_text.txt --fontconfig_tmpdir=/tmp/font_tmp.KohYfRN1r2
Rendered page 0 to file /tmp/font_tmp.KohYfRN1r2/sample_text.txt.tif

=== Phase I: Generating training images ===
Rendering using Arial
[Fri 16 Aug 00:04:32 BST 2019] /usr/bin/text2image --fontconfig_tmpdir=/tmp/font_tmp.KohYfRN1r2 --fonts_dir=/usr/share/fonts --strip_unrenderable_words --leading=32 --xsize=3600 --char_spacing=0.0 --exposure=0 --outputbase=/tmp/eng-2019-08-16.c0r/eng.Arial.exp0 --max_pages=0 --font=Arial --text=../langdata_lstm/eng/eng.training_text
Rendered page 0 to file /tmp/eng-2019-08-16.c0r/eng.Arial.exp0.tif
Rendered page 1 to file /tmp/eng-2019-08-16.c0r/eng.Arial.exp0.tif
Rendered page 2 to file /tmp/eng-2019-08-16.c0r/eng.Arial.exp0.tif
Rendered page 3 to file /tmp/eng-2019-08-16.c0r/eng.Arial.exp0.tif
Rendered page 4 to file /tmp/eng-2019-08-16.c0r/eng.Arial.exp0.tif
Rendered page 5 to file /tmp/eng-2019-08-16.c0r/eng.Arial.exp0.tif
Rendered page 6 to file /tmp/eng-2019-08-16.c0r/eng.Arial.exp0.tif
Rendered page 7 to file /tmp/eng-2019-08-16.c0r/eng.Arial.exp0.tif
Rendered page 8 to file /tmp/eng-2019-08-16.c0r/eng.Arial.exp0.tif
Rendered page 9 to file /tmp/eng-2019-08-16.c0r/eng.Arial.exp0.tif

=== Phase UP: Generating unicharset and unichar properties files ===
[Fri 16 Aug 00:04:38 BST 2019] /usr/bin/unicharset_extractor --output_unicharset /tmp/eng-2019-08-16.c0r/eng.unicharset --norm_mode 1 /tmp/eng-2019-08-16.c0r/eng.Arial.exp0.box
Extracting unicharset from box file /tmp/eng-2019-08-16.c0r/eng.Arial.exp0.box
Other case É of é is not in unicharset
Wrote unicharset file /tmp/eng-2019-08-16.c0r/eng.unicharset
[Fri 16 Aug 00:04:38 BST 2019] /usr/bin/set_unicharset_properties -U /tmp/eng-2019-08-16.c0r/eng.unicharset -O /tmp/eng-2019-08-16.c0r/eng.unicharset -X /tmp/eng-2019-08-16.c0r/eng.xheights --script_dir=../langdata_lstm
Loaded unicharset of size 112 from file /tmp/eng-2019-08-16.c0r/eng.unicharset
Setting unichar properties
Other case É of é is not in unicharset
Setting script properties
Warning: properties incomplete for index 78 = ~
Writing unicharset to file /tmp/eng-2019-08-16.c0r/eng.unicharset

=== Phase E: Generating lstmf files ===
Using TESSDATA_PREFIX=./tessdata/eng
[Fri 16 Aug 00:04:38 BST 2019] /usr/bin/tesseract /tmp/eng-2019-08-16.c0r/eng.Arial.exp0.tif /tmp/eng-2019-08-16.c0r/eng.Arial.exp0 --psm 6 lstm.train
Error opening data file ./tessdata/eng/eng.traineddata
Please make sure the TESSDATA_PREFIX environment variable is set to your "tessdata" directory.
Failed loading language 'eng'
Tesseract couldn't load any languages!
Could not initialize tesseract.
ERROR: Program tesseract failed. Abort.
erdos:/tmp/tesstrain $ 
```

Note that the error here is that it couldn't find the eng.traineddata file.

Best wishes,

Julian

**Comment by amitdo** [comment]
>Here's the output of running tesstrain.sh without a traineddata file already prepared:

The traineddata should be from the 'traineddata_best' repo.

**Comment by Shreeshrii** [comment]
>>following the wiki instructions, use combine_lang_model to create a minimal traineddata file, then use tesstrain.sh to create initial training data before training a new neural network to perform the OCR. 

I think the wiki instructions need to be updated.

tesstrain.sh takes care of building the minimal/starter trained data since it runs the combine_lang_model command as one of the steps.

It needs to be run as a separate step for someone who may be training using their own script eg. For training from scanned image box/tiff files rather than synthetic data from training text and fonts.

**Comment by Shreeshrii** [comment]
>>that it couldn't find the eng.traineddata file.

I think this is related to tesseract initialization which requires both osd and eng traineddata ti be available in tessdata_prefix path.

**Comment by juliangilbey** [comment]
In response to several of the above comments:

> > > that it couldn't find the eng.traineddata file.
> 
> I think this is related to tesseract initialization which requires both osd and eng traineddata ti be available in tessdata_prefix path.

That is correct, and that is why the wiki page explains that to train from scratch, one needs to create starter traineddata using the combine_lang_model program.  (tesseract does not seem to require osd traineddata when generating the initial LSTM training data, though.)

> tesstrain.sh takes care of building the minimal/starter trained data since it runs the combine_lang_model command as one of the steps.

Yes and no.  It does run this in the make__lstmdata function, as I said in an earlier comment, but that runs *after* the phase_E_extract_features function.  And since the phase_E_extract_features function calls tesseract, it requires a traineddata file to already be present before running tesstrain.sh.  The call to combine_lang_model that you mention appears to overwrite the initial traineddata file (though I haven't investigated this step in depth), but it does not help with this bug.

> > Here's the output of running tesstrain.sh without a traineddata file already prepared:
> 
> The traineddata should be from the 'traineddata_best' repo.

Using pre-existing trained data would defeat the point of "training from scratch".

Looking in the code, in ccmain/tessedit.cpp, lines 161-165, we see:

```
  // If we are only loading the config file (and so not planning on doing any
  // recognition) then there's nothing else do here.
  if (tessedit_init_config_only) {
    return true;
  }
```

This is exactly the situation we are in, and hence why my proposed patch works.

Note that the wiki *does* need to be updated, but only to say that (the fixed) lstm.train needs to be copied to tessdata/eng/configs.

Best wishes,

Julian

**Comment by amitdo** [comment]
>Using pre-existing trained data would defeat the point of "training from scratch"

I don't think so. Did you actually try to do so?

You can replace some of the files in langdata with your own files, and tesstrain will use them to build the traineddata.

Some parts of the wiki explain how do training completely manually, **without using tesstrain.sh**.

**Comment by juliangilbey** [comment]
I wonder whether a parable might help to clarify some misunderstandings or confusions here?

Reader: Hi, I've been using your excellent recipe book and trying to make your Victoria Sponge.  But it's really not worked well - something went wrong....

Reader: Ah, I've figured it - somehow the instruction to add sugar to the batter was left out in step 2.  It's in the list of ingredients, but it has forgotten to say when to add it.

Author: But it says add sugar in step 4?  I don't understand the problem.

Reader: Yes, indeed, you want to sprinkle icing sugar over the cake in step 4.  But if you follow the recipe as written, it goes wrong because it doesn't say to add sugar to the batter in step 2.

Author: Anyway, I don't understand why you are trying to make a Victoria Sponge when you can make a chocolate cake.  There are instructions for making a very tasty chocolate cake on page 73!

Reader: But I want to make a Victoria Sponge!  And you have a really good Victoria Sponge recipe in the book, and there's just one small issue with it, which is easily remedied, as I've explained!

Author: I really don't understand you.  The chocolate cake is so much better!!

**Comment by Shreeshrii** [comment]
@juliangilbey You are following your own recipe and not really following instructions in the wiki.

If you want to train from scratch using tesstrain.sh, the instructions are:

```
src/training/tesstrain.sh --fonts_dir /usr/share/fonts --lang eng --linedata_only \
  --noextract_font_properties --langdata_dir ../langdata \
  --tessdata_dir ./tessdata --output_dir ~/tesstutorial/engtrain
```

```
src/training/tesstrain.sh --fonts_dir /usr/share/fonts --lang eng --linedata_only \
  --noextract_font_properties --langdata_dir ../langdata \
  --tessdata_dir ./tessdata \
  --fontlist "Impact Condensed" --output_dir ~/tesstutorial/engeval
```

```
mkdir -p ~/tesstutorial/engoutput
training/lstmtraining --debug_interval 1 \
  --traineddata ~/tesstutorial/engtrain/eng/eng.traineddata \
  --net_spec '[1,36,0,1 Ct3,3,16 Mp3,3 Lfys48 Lfx96 Lrx96 Lfx256 O1c111]' \
  --model_output ~/tesstutorial/engoutput/base --learning_rate 20e-4 \
  --train_listfile ~/tesstutorial/engtrain/eng.training_files.txt \
  --eval_listfile ~/tesstutorial/engeval/eng.training_files.txt \
  --max_iterations 500
```

Please give these a try. You don't need to run combine_lang_model separately.

**Comment by Shreeshrii** [comment]
You can also clone https://github.com/Shreeshrii/tess4training and run the scripts in that to get a feel of tesstutorial and LSTM training process.

**Comment by amitdo** [comment]
Julian, forget the recipe for now (you misread it) and listen to the chefs :-)

**Comment by juliangilbey** [comment]
Hi @Shreeshrii, and thanks for your message.

I think I now see where the confusion lies.

The section entitled [Creating Starter Traineddata](https://github.com/tesseract-ocr/tesseract/wiki/TrainingTesseract-4.00#creating-starter-traineddata) is in the section on lstmtraining, and I hadn't appreciated that the intention of the wiki was that this only applied to the lstmtraining step.  But to create the training data (using tesstrain.sh or the other method described), one needs to already have a traineddata file, hence the need to download it from the repository before running tesstrain.sh.

However, it turns out that one can use exactly the same starter traineddata file to run tesstrain.sh; .  This is on condition that lstm.train contains `tessedit_init_config_only T`, so that no attempt is made to load the TESSDATA_LSTM component.  Presumably, that's what the tesseract developers did when they did the very first training, before there was an existing traineddata file containing a TESSDATA_LSTM component.  I wonder what would happen if someone wanted to train for a new language for which there doesn't exist a traineddata file yet?  Would they be able to use an English traineddata file to create the lstmf box files for this new language?  (And the lstmbox approach described on the wiki also requires an existing traineddata file.)  I don't know enough about tesseract yet to know the answer.

Since `tessedit_init_config_only T` causes no harm (the TESSDATA_LSTM component is not used in the box creation code), it would seem fine to add it.

Does that make a little more sense?
